### ISSUE TITLE
Data loss prevention with jsbin.com

### ISSUE THREAD

**Issue by sfbwalder** [jsbin/jsbin#2410]
**The problem...**  Our corporation is very concerned about data loss prevention.  Access to jsbin.com has been blocked from the corporate network.  This is primarily due to the risk of jsbin.com being used as a conduit to move files out to drop-box, but a secondary concern is being able to save as in a directory structure or html file.  

**The question...**  As developers, we would really like to have access to jsbin.com, but we need to have a solution to mitigate this risk.  Has anyone else had this issue, and if so, could you please share potential solutions?

**More thoughts...**  We haven't looked closely at the source, hoping someone could suggest a feature quickly:
* Having a pro account provides access to these features.  Is there a way to prevent logging into a pro account by blocking a specific url rather than the entire domain?  If not, is there a way to prevent the save/transfer by blocking a specific url?  
* Is there a suggested feature we could contribute to jsbin.com that would help us and other corporations who may have the same concern?
* Running a copy internally doesn't seem to give us anything, as we want/need to share with the development community.
* Other thoughts?

**The scoffing...**  As developers, we sometimes scoff at this security policy, and push the question  "why bother blocking jsbin.com when there are probably a million other avenues for data loss?"  All that is beside the point at this time.  When corporate security / law decides a site is a risk, it is blocked, and the risk has to appropriately mitigated before it will be opened.

**Comment by remy** [comment]
I think it's worth fully understanding the problems you're referencing before discussing suggestions (most of which either don't fly, like the local install, or can be worked around by a user).

Anyway, the two items:

1. being used as a conduit to move files out to drop-box
2. being able to save as in a directory structure or html file

## Dropbox

(For pro users) JS Bin can *only* save the bin's HTML to a single directory that's determined by dropbox's API. 

The only thing that can happen here (and I put the context as someone in your company), is the output of a page is saved to a dropbox specific directory.

Without that feature, *anyone* can view source on the full output and save *that* to a folder.

## Saving to a directory structure

Assuming this is different from saving to dropbox (and again, dropbox is to a single specific directory that dropbox's api defined), this isn't currently possible in jsbin. You can download the bin, but *you* the user decide where it goes. And again, you can view source and save, but that's it.

This is a limitation of the browser's native security, nothing else.

---

Can you clarify if I've got that right or wrong?

**Comment by sfbwalder** [comment]
We may not be understanding the drop box integration.  We don't have a Pro account yet, as we are working through purchasing and corp law to pay for that, so we are going off what we have inferred from reading through the various capabilities and issues.  So, we apologize for our current lack of ability to just see it first hand.

**Here are the presumed exposures**
*  If I can upload an asset (a file of private company info) to a bin, and then save the contents of the bin to my personal dropbox folder, and then share my dropbox folder with a information broker, that seems like a conduit for data leakage.
 * When you say "can only download the bins html", we may be misunderstanding what that means.
 * We may be misunderstanding what the upload of an asset means also.
* Same thing for saving to a directory.  I can upload an asset from my work pc, then access my jsbin.com pro account from home, and then download the bin contents (and asset) to my local pc, etc.
 * When you say "can only download the bin", we may be misunderstanding what that capabilities are.

Perhaps it would be best if we got a Pro account for a month and tried it out so we could have a more educated dialog from our side.  If you give out trial pro accounts for a month, we would use one, but since you run on donations we don't expect that from you.  I will continue to work through our (slow) corporate processes needed to make a cc payment on the site. 

Thanks, Ben

**Comment by remy** [comment]
This all sounds like you have, generally, the right understanding. If you upload an asset, this isn't synced to a dropbox, but the bin containing the URL to the asset would be in your dropbox, and you're right, you could go home and visit the URL to the asset. Basically it's the same as emailing yourself a file in this case - the process isn't the same, but the outcome is.

This is what I'm referring to by "download": 

![screen shot 2015-06-10 at 16 51 38](https://cloud.githubusercontent.com/assets/13700/8087140/183d8b92-0f91-11e5-9a63-1cdcc29857d4.png)


I'm not sure I fully understand this extremely restrictive security policy, but I assume you're also not allow to use email or POST forms - since the same security exposure occurs. 

And I understand if the data is sensitive, you don't want it posted on a network like the internet.

i.e. you can post private company information in *this* form (though I'm guessing that the company hasn't blocked github.com yet).

Bottom line, going by what you've described, and according to the restrictions you've explained, I think jsbin might not work inside the policy.

**Comment by sfbwalder** [comment]
You hit at the heart of our issue... and the scoffing this gets from most developers...
> I'm not sure I fully understand this extremely restrictive security policy, but I assume you're also not allow to use email or POST forms - since the same security exposure occurs.
>And I understand if the data is sensitive, you don't want it posted on a network like the internet.
>i.e. you can post private company information in this form (though I'm guessing that the company hasn't blocked github.com yet).

We do use email, and we can post to millions of sites.  But, we still have to prove this one is has no more risk than anywhere else, or attempt to come up with a solution to mitigate the risk.

We will pursue our process to pay for a pro account, and then we will be able to better assess how this works.

I am still hoping some other corporation similar to ours (banking, insurance) might have come up with a solution to open up access to just part of the site by being able to block the features that are a concern, but still have access to all the examples etc. that jsbin.com is used for.

**Comment by remy** [comment]
Like I said in my earlier comments, I'm really not sure there's *anything* that can be done at any technical level that can prevent someone from copying and pasting (or straight download - which is just a convenience function).

I'm open to ideas if it helps, but I'll close this issue for now and hope you're able to get somewhere with your companies policies.

Good luck!
